### PR TITLE
Add 2015-10-27-csharp-interactive

### PR DESCRIPTION
**Pull request by terrajobst**

### CODE REVIEW

File: 2015-10-27-csharp-interactive/APIs.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  }
+ }
+ namespace Microsoft.CodeAnalysis.CSharp.Scripting.Hosting {
+  public sealed class CSharpObjectFormatter : ObjectFormatter {
+    public static CSharpObjectFormatter Instance { get; }
+  }
+ }
+ namespace Microsoft.CodeAnalysis.Scripting {
+  public sealed class CompilationErrorException : Exception {
+    ^ Immo Landwerth: Should have public .ctors
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Done.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    public ImmutableArray<Diagnostic> Diagnostics { get; }
+  }
+  public abstract class Script {
+    public string Code { get; }
+    public Type GlobalsType { get; }
+    public ScriptOptions Options { get; }
+    public Script Previous { get; }
+    public abstract Type ReturnType { get; }
+    public ImmutableArray<Diagnostic> Build(CancellationToken cancellationToken=null);
+    ^ Immo Landwerth: Should be called Compile
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Done.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    public Script<object> ContinueWith(string code, ScriptOptions options=null);
+    public Script<TResult> ContinueWith<TResult>(string code, ScriptOptions options=null);
+    public Compilation GetCompilation();
+    public Task<ScriptState> RunAsync(object globals=null, CancellationToken cancellationToken=null);
+    public Script WithCode(string code);
+    ^ Immo Landwerth: Remove. Confusion with ContinueWith.
+    public Script WithGlobalsType(Type globalsType);
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Removed.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    ^ Immo Landwerth: Remove. Unclear what the chaining semantics are.
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Removed.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    public Script WithOptions(ScriptOptions options);
+  }
+  public sealed class Script<T> : Script {
+    public override Type ReturnType { get; }
+    public ScriptRunner<T> CreateDelegate(CancellationToken cancellationToken=null);
+    public new Task<ScriptState<T>> RunAsync(object globals=null, CancellationToken cancellationToken=null);
+    public new Script<T> WithCode(string code);
+    ^ Immo Landwerth: Same
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Removed.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    public new Script<T> WithGlobalsType(Type globalsType);
+    ^ Immo Landwerth: Same
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Removed.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    public override PortableExecutableReference ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity);
+    public override ImmutableArray<PortableExecutableReference> ResolveReference(string reference, string baseFilePath, MetadataReferenceProperties properties);
+    public ScriptMetadataResolver WithBaseDirectory(string baseDirectory);
+    public ScriptMetadataResolver WithSearchPaths(IEnumerable<string> searchPaths);
+    public ScriptMetadataResolver WithSearchPaths(ImmutableArray<string> searchPaths);
+    public ScriptMetadataResolver WithSearchPaths(params string[] searchPaths);
+  }
+  public sealed class ScriptOptions {
+    public static readonly ScriptOptions Default;
+    ^ Immo Landwerth: Make a property
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Done.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    public ScriptOptions WithReferences(IEnumerable<MetadataReference> references);
+    public ScriptOptions WithReferences(IEnumerable<Assembly> references);
+    public ScriptOptions WithReferences(IEnumerable<string> references);
+    public ScriptOptions WithReferences(params Assembly[] references);
+    public ScriptOptions WithReferences(params string[] references);
+    public ScriptOptions WithSourceResolver(SourceReferenceResolver resolver);
+  }
+  public delegate Task<T> ScriptRunner<T>(object globals=null, CancellationToken cancellationToken=null);
+  public sealed class ScriptSourceResolver : SourceFileResolver, IEquatable<ScriptSourceResolver> {
+    ^ Immo Landwerth: You should consider having == because IEquatable<>. Perhaps you want to offload Equals into a separate comparer.
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: The scenario where equality is needed is when CompilationOptions are compared (CSharp and VB compilation options implement IEquatable). CompilationOtpions have SourceReferenceResolver and MetadataReferenceResolver fields and CompilationOptions.Equals calls Equals on these resolvers. ScriptSourceResolver derives from SourceReferenceResolver and overrides Equals. 

None of the existing (already shipped) classes implement operators == and !=. So I'm going to keep it consistent for now.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    public ScriptSourceResolver WithSearchPaths(params string[] searchPaths);
+  }
+  public abstract class ScriptState {
+    public object ReturnValue { get; }
+    public Script Script { get; }
+    public ImmutableArray<ScriptVariable> Variables { get; }
+    public Task<ScriptState<object>> ContinueWithAsync(string code, ScriptOptions options=null, CancellationToken cancellationToken=null);
+    public Task<ScriptState<TResult>> ContinueWithAsync<TResult>(string code, ScriptOptions options=null, CancellationToken cancellationToken=null);
+    public ScriptVariable GetVariable(string name);
+    ^ Immo Landwerth: LookupVariable()? This conveys language specific lookup rules around hiding. Should throw if name doesn't exist at all. If you don't want to throw TryLookupVariable(name, out ScriptVariable)
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Changed to TryLookupVariable

File: 2015-10-27-csharp-interactive/APIs.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  }
+  public sealed class ScriptState<T> : ScriptState {
+    public new T ReturnValue { get; }
+  }
+  public sealed class ScriptVariable {
+    ^ Immo Landwerth: We should probably expose whether the variable is read-only.
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Added bool IsReadOnly { get; }

File: 2015-10-27-csharp-interactive/APIs.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    public Metadata GetMetadata(string fullPath, MetadataImageKind kind);
+    public MetadataShadowCopy GetMetadataShadowCopy(string fullPath, MetadataImageKind kind);
+    public bool IsShadowCopy(string fullPath);
+    public bool NeedsShadowCopy(string fullPath);
+    public void SuppressShadowCopy(string originalPath);
+  }
+  public abstract class ObjectFormatter {
+    public string FormatObject(object obj);
+  }
+  public sealed class ShadowCopy {
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Renamed to ShadowCopyFile.

File: 2015-10-27-csharp-interactive/APIs.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    public void RegisterDependency(Assembly dependency);
+  }
+  public class InteractiveScriptGlobals {
+    public InteractiveScriptGlobals(TextWriter outputWriter, ObjectFormatter objectFormatter);
+    public IList<string> Args { get; }
+    public IList<string> ReferencePaths { get; }
+    public IList<string> SourcePaths { get; }
+    public void Print(object value);
+  }
+  public sealed class MetadataShadowCopy {
```
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Renamed

File: 2015-10-27-csharp-interactive/APIs.md (hunk 11 above)
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Renamed to FileShadowCopy.

File: 2015-10-27-csharp-interactive/APIs.md (hunk 8 above)
Comment (terrajobst) [inline 2015-10-27-csharp-interactive/APIs.md]: Makes sense. Thanks for following up!

File: 2015-10-27-csharp-interactive/APIs.md (hunk 9 above)
Comment (terrajobst) [inline 2015-10-27-csharp-interactive/APIs.md]: Good choice :-)

File: 2015-10-27-csharp-interactive/APIs.md (hunk 10 above)
Comment (terrajobst) [inline 2015-10-27-csharp-interactive/APIs.md]: Great.

File: 2015-10-27-csharp-interactive/APIs.md (hunk 9 above)
Comment (tmat) [inline 2015-10-27-csharp-interactive/APIs.md]: Actually, I reverted the change. MattWar pointed out that we do it in other Roslyn APIs and it's just simpler to return null.

Comment (terrajobst) [inline 2015-10-27-csharp-interactive/APIs.md]: Works for me.

**Pull request merged**